operator/duties/scheduler.go

```diff
@@ -0,0 +1,342 @@
+						zap.String("new_previous_dependent_root", fmt.Sprintf("%#x", data.PreviousDutyDependentRoot[:])))
+
+					s.reorg <- ReorgEvent{
+						Slot:     data.Slot,
+						Previous: true,
+					}
+				}
+
+				// Ensure that the current dependent roots are the same.
+				if !bytes.Equal(s.currentDutyDependentRoot[:], zeroRoot[:]) &&
+					!bytes.Equal(s.currentDutyDependentRoot[:], data.CurrentDutyDependentRoot[:]) {
+					logger.Debug("🔀 Current duty dependent root has changed",
+						zap.String("old_current_dependent_root", fmt.Sprintf("%#x", s.currentDutyDependentRoot[:])),
+						zap.String("new_current_dependent_root", fmt.Sprintf("%#x", data.CurrentDutyDependentRoot[:])))
+
+					s.reorg <- ReorgEvent{
+						Slot:    data.Slot,
+						Current: true,
+					}
+				}
+			}
+		}
+
+		s.lastBlockEpoch = epoch
+		s.previousDutyDependentRoot = data.PreviousDutyDependentRoot
+		s.currentDutyDependentRoot = data.CurrentDutyDependentRoot
+
+		currentTime := time.Now()
+		delay := s.network.SlotDurationSec() / 3 /* a third of the slot duration */
+		slotStartTimeWithDelay := s.network.Beacon.GetSlotStartTime(data.Slot).Add(delay)
+		if currentTime.Before(slotStartTimeWithDelay) {
+			logger.Debug("🏁 Head event: Block arrived before 1/3 slot", zap.Duration("time_saved", slotStartTimeWithDelay.Sub(currentTime)))
+
+			// We give the block some time to propagate around the rest of the
+			// nodes before kicking off duties for the block's slot.
+			time.Sleep(s.blockPropagateDelay)
+
+			s.waitCond.L.Lock()
+			s.headSlot = data.Slot
+			s.waitCond.Broadcast()
+			s.waitCond.L.Unlock()
+		}
+	}
+}
+
+// ExecuteDuties tries to execute the given duties
+func (s *Scheduler) ExecuteDuties(logger *zap.Logger, duties []*spectypes.Duty) {
+	for _, duty := range duties {
+		duty := duty
+		logger := s.loggerWithDutyContext(logger, duty)
+		go func() {
+			if duty.Type == spectypes.BNRoleAttester || duty.Type == spectypes.BNRoleSyncCommittee {
+				s.waitOneThirdOrValidBlock(duty.Slot)
+			}
+			s.executeDuty(logger, duty)
+		}()
+	}
+}
+
+// loggerWithDutyContext returns an instance of logger with the given duty's information
+func (s *Scheduler) loggerWithDutyContext(logger *zap.Logger, duty *spectypes.Duty) *zap.Logger {
+	return logger.
+		With(fields.Role(duty.Type)).
+		With(zap.Uint64("committee_index", uint64(duty.CommitteeIndex))).
+		With(fields.CurrentSlot(s.network.Beacon.EstimatedCurrentSlot())).
+		With(fields.Slot(duty.Slot)).
+		With(fields.Epoch(s.network.Beacon.EstimatedEpochAtSlot(duty.Slot))).
+		With(fields.PubKey(duty.PubKey[:])).
+		With(fields.StartTimeUnixMilli(s.network.Beacon.GetSlotStartTime(duty.Slot)))
+}
+
+// waitOneThirdOrValidBlock waits until one-third of the slot has transpired (SECONDS_PER_SLOT / 3 seconds after the start of slot)
+func (s *Scheduler) waitOneThirdOrValidBlock(slot phase0.Slot) {
+	// Wait for the event or signal
```

Comment:
aren't we waiting only for the block here?

Reply:
no, we are waiting for the block from head event or for the one third of the slot release from scheduler slot ticker